src/classes/region.h

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Team Dissolve and contributors
+
+#pragma once
+
+#include "classes/box.h"
+#include "classes/configuration.h"
+#include "templates/array3d.h"
+#include "templates/vector3.h"
+#include <vector>
+
+// Forward Declarations
+class Configuration;
+
+// Region Data
+class Region
+{
+    public:
+    Region();
+    ~Region() = default;
+
+    private:
+    // Box from target Configuration
+    const Box *box_;
+    // 3D map of available voxels
+    Array3D<std::pair<Vec3<int>, bool>> voxelMap_;
+    // Number of voxels along each cell axis
+    Vec3<int> nVoxels_;
+    // Fractional voxel size
+    Vec3<double> voxelSizeFrac_;
+    // Lower-left corner voxel indices of free regions
+    std::vector<std::pair<Vec3<int>, bool>> freeVoxels_;
+
+    public:
+    // Generate region information
+    template <class Lam> bool generate(const Configuration *cfg, double voxelSize, Lam voxelCheckFunction)
+    {
+        box_ = cfg->box();
+
+        // Set fractional voxel sizes
+        for (auto n = 0; n < 3; ++n)
+            nVoxels_.set(n, std::max(int(box_->axisLength(n) / voxelSize), 1));
+        voxelSizeFrac_.set(1.0 / nVoxels_.x, 1.0 / nVoxels_.y, 1.0 / nVoxels_.z);
+
+        // Initialise 3D map and determine valid voxels
+        voxelMap_.initialise(nVoxels_.x, nVoxels_.y, nVoxels_.z);
+        for (auto x = 0; x < nVoxels_.x; ++x)
+            for (auto y = 0; y < nVoxels_.y; ++y)
+                for (auto z = 0; z < nVoxels_.z; ++z)
+                    voxelMap_[{x, y, z}] = {
+                        Vec3<int>(x, y, z),
+                        voxelCheckFunction(cfg, box_->getReal({(x + 0.5) * voxelSizeFrac_.x, (y + 0.5) * voxelSizeFrac_.y,
+                                                               (z + 0.5) * voxelSizeFrac_.z}))};
+
+        // Create linear vector of all available voxels
+        auto nFreeVoxels = std::count_if(voxelMap_.begin(), voxelMap_.end(), [](const auto &voxel) { return voxel.second; });
+        freeVoxels_.clear();
+        freeVoxels_.resize(nFreeVoxels);
+        std::copy_if(voxelMap_.begin(), voxelMap_.end(), freeVoxels_.begin(), [](const auto &voxel) { return voxel.second; });
```

Comment:
```suggestion
        auto nFreeVoxels = std::count_if(voxelMap_.begin(), voxelMap_.end(), std::get<1>);
        freeVoxels_.clear();
        freeVoxels_.resize(nFreeVoxels);
        std::copy_if(voxelMap_.begin(), voxelMap_.end(), freeVoxels_.begin(), std::get<1>);
```

No need to define a new lambda.

Reply:
This can't work because `std::get<>` needs the `std::pair` object as a parameter, unless I'm missing some C++ voodoo in which the current item can be implicitly passed?

Reply:
`std::get<1>` should get the object as a parameter just as the lambda function is passed the object as a parameter, unless I'm missing something?

Reply:
Sounds sensible, but doesn't compile. Not in C++17 perhaps?

Reply:
I learned something new today:  taking function pointers to functions in the `std` namespace is explicitly defined in the standard as being undefined behaviour.  I've rolled back to before I made the suggestion.